src/components/NumberInput/NumberInput.js

```diff
@@ -0,0 +1,198 @@
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import Icon from '../Icon';
+import classNames from 'classnames';
+
+export default class NumberInput extends Component {
+  static propTypes = {
+    className: PropTypes.string,
+    disabled: PropTypes.bool,
+    iconDescription: PropTypes.string.isRequired,
+    id: PropTypes.string.isRequired,
+    label: PropTypes.string,
+    max: PropTypes.number,
+    min: PropTypes.number,
+    onChange: PropTypes.func,
+    onClick: PropTypes.func,
+    value: PropTypes.number,
+    step: PropTypes.number,
+    invalid: PropTypes.bool,
+    invalidText: PropTypes.string,
+  };
+
+  static defaultProps = {
+    disabled: false,
+    iconDescription: 'choose a number',
+    label: ' ',
+    onChange: () => {},
+    onClick: () => {},
+    value: 0,
+    step: 1,
+    invalid: false,
+    invalidText: 'Provide invalidText',
+  };
+
+  /**
+   * The DOM node refernce to the `<input>`.
+   * @type {HTMLInputElement}
+   */
+  _inputRef = null;
+
+  constructor(props) {
+    super(props);
+
+    let value = props.value;
+    if (props.min || props.min === 0) {
+      value = Math.max(props.min, value);
+    }
+
+    this.state = {
+      value,
+      labelMotion:
+        !!props.defaultValue || !!props.value || props.defaultValue != null || props.value != null,
```

Comment:
Could we branch this into its own component soon? Seeing this pop up for textInput etc too, right?